synchrona: add Status LEDs control

tachicialex opened the pull request:
Create a subprocess that handles the LED color based
on the Locking Status of AD9545 and HMC7044.

Signed-off-by: Alexandru Tachici <alexandru.tachici@analog.com>

nunojsa commented on app/python/synchrona/ad-synchrona-14-leds.sh:
```diff
@@ -0,0 +1,94 @@
+
+#export all led gpios and set as output
+echo $RED_LED1 > /sys/class/gpio/export
+echo "out" > "/sys/class/gpio/gpio$RED_LED1/direction"
```
nit: this does not hurt but these pins should already be set to out by the pinctrl... Note that this can actually be obfuscating a bug in the devicetree. That was the case for the vcxo selection

nunojsa commented on app/python/synchrona/ad-synchrona-14-leds.sh:
```diff
@@ -0,0 +1,94 @@
+	if [[ $LED_STATUS == $STATUS_GREEN ]]; then
+		echo 0 > "/sys/class/gpio/gpio$RED_LED1/value"
+		echo 1 > "/sys/class/gpio/gpio$GREEN_LED1/value"
+		sleep 1
```
no sleep in between the toggling?

tachicialex replied:
There are two leds behind the same transparent screen. no need to to sleep here.

nunojsa commented on app/python/synchrona/ad-synchrona-14-leds.sh:
```diff
@@ -0,0 +1,94 @@
+		echo 1 > "/sys/class/gpio/gpio$RED_LED1/value"
+		echo 0 > "/sys/class/gpio/gpio$GREEN_LED1/value"
+		sleep 1
+	fi
```
I can see that there are some differences in the values used for the sleep, but can't we have just sleep 1 before going to the next iteration of the loop? Note that I'm not speaking about the sleeps between toggling the same pin...

nunojsa commented on app/python/synchrona/ad-synchrona-14-leds.sh:
```diff
@@ -0,0 +1,94 @@
+		echo 0 > "/sys/class/gpio/gpio$GREEN_LED1/value"
+		sleep 1
+	fi
+done
```
always a good police to have -e on a bash script

tachicialex replied:
Thought about that, but process will exit when it won't find HMC7044_STATUS_PATH during a reload of the overlay. In this way, in case of error, will keep the LED red.

nunojsa commented on app/python/synchrona/main.py:
```diff
@@ -57,6 +58,11 @@
+os.system("ps axf | grep \"ad-synchrona-14-leds\.sh\" | grep -v grep | awk '{print \"kill -9 \" $1}' | sh")
+subprocess.Popen(['bash', 'ad-synchrona-14-leds.sh'], stderr = open(os.devnull, 'w'),
+    stdout = open(os.devnull, 'w'))
+
```
This looks very odd and hacky to me... Is there any reason why we want to kill the script and start it again every time we get in here?

The way I see it, this script looks like it's something that should be started/stopped by the synchrona systemd service...

tachicialex commented:
Changelog v1 -  v2:
   - in _setup_synchrona_service(): added ExecStartPost for starting the LEDs script, no longer
   start from app/python/synchrona/main.py
   - sleep 1 sec in while()
   - don't set the output direction

nunojsa commented on installers/common.sh:
```diff
@@ -207,6 +207,7 @@ function _setup_synchrona_service() {
     ExecStartPre=-/usr/bin/dtoverlay /boot/overlays/rpi-ad9545-hmc7044.dtbo
     ExecStartPre=-/usr/bin/sh /var/www/html/app/python/synchrona/rebind.sh
     ExecStart=/usr/local/bin/uvicorn --app-dir=/var/www/html/app/python/synchrona  main:app --host 0.0.0.0 --port 8000
+    ExecStartPost=-/usr/bin/bash /var/www/html/app/python/synchrona/ad-synchrona-14-leds.sh > /dev/null 2>&1 &
```
nope... we should not be doing this like this. You should not hardcode `/var/www/html/` as the user might choose a different installation path. I do have a fix for this in [here](https://github.com/analogdevicesinc/raspap-webgui/pull/2/commits/1472a9c51273dd67b2ca8536a7f42fada704bf88). It will be merged in the next few days...

I would actually move your script to installers/ and during installation move it to /etc/raspap/synchrona since, AFAICT, that path is not configurable by the user.

tachicialex commented:
Changelog v2 -> v3:
- moved script to installers/

nunojsa commented:
Maybe it would be a good idea to rebase your branch...

tachicialex commented:
This is already rebased over [synchrona_rev](https://github.com/analogdevicesinc/raspap-webgui/tree/synchrona_revB), HEAD~1 is 9f38e1fadbbe. Is this the right branch?

nunojsa commented on installers/common.sh:
```diff
@@ -208,6 +208,7 @@ function _setup_synchrona_service() {
     ExecStartPre=-/usr/bin/dtoverlay /boot/overlays/rpi-ad9545-hmc7044.dtbo
     ExecStart=/usr/local/bin/uvicorn --app-dir=$tmp/app/python/synchrona  main:app --host 0.0.0.0 --port 8000
     ExecStopPost=-/usr/bin/sh -c "/usr/bin/echo 6 > /sys/class/gpio/unexport"
+    ExecStartPost=-/usr/bin/bash /etc/raspap/synchrona/ad-synchrona-14-leds.sh > /dev/null 2>&1 &
```
Did you make sure the script is killed when you do "systemctl stop synchrona"? At first glance, I don't think it is... Also, as a nit comment, add this line after `ExecStart`as this is part of the service startup.

At this point, I'm not sure if the LED script should not have it's own systemd service that could automatically start when one starts the synchrona service. I'm fairly sure this is doable but not 100% sure though...

Anyways, maybe just make sure you kill the script when the service is stopped for now...

tachicialex replied:
So I checked and the script gets killed on: "service synchrona stop"

tachicialex commented:
Changelog v3 -> v4:
- moved ExecStartPost after ExecStart
- Added export/unexport commands for gpio 12 and 16 (apparently the export; DT property for pincfg does not do anything)

nunojsa commented on installers/common.sh:
```diff
@@ -739,4 +745,3 @@ function _install_complete() {
         fi
     fi
 }
-
```
Hmm I guess your editor is doing something :smile: ...

tachicialex committed:
synchrona: add Status LEDs control

Create a subprocess that handles the LED color based
on the Locking Status of AD9545 and HMC7044.

Signed-off-by: Alexandru Tachici <alexandru.tachici@analog.com>

tachicialex commented:
Changelog v4 -> v5:
- fixed whatever was that thing my Atom did

tachicialex commented:
Changelog v5 -> v6:
- set in script the GPIO direction

nunojsa commented on installers/ad-synchrona-14-leds.sh:
```diff
@@ -0,0 +1,83 @@
+while true
+do
+	ad9545_pll_status=$(cat $AD9545_STATUS_PATH)
+	hmc7044_status=$(cat $HMC7044_STATUS_PATH)
```
nit: just realized this now.... You could  use libiio tools to read the status and so you would not need to make assumptions on the device path. The whole thing already depends on libiio anyways so I think it's fine to assume it is installed. No need to do the change though

tachicialex commented:
Changelog v5 -> v6:
- rebased again over synchrona_revB
